File: 5. LinkedLists/5. reverse_list/reverse_list.js
// class Node {
//   constructor(val) {
//     this.val = val;
//     this.next = null;
//   }
// }

// const a = new Node("a");
// const b = new Node("b");
// const c = new Node("c");
// const d = new Node("d");
// const e = new Node("e");
// const f = new Node("f");

// a.next = b;
// b.next = c;
// c.next = d;
// d.next = e;
// e.next = f;

// Iterative
const reverseList = (head) => {
    let prev = null;
    let current = head;
    while(current !== null){
        let next = current.next;
        current.next = prev;
        prev = current;
        current = next;
    }
    return prev;
}

// Recursive
// const reverseList = (head, prev=null) => {
//     if(head === null) return prev;
//     let next = head.next;
//     head.next = prev;
//     return reverseList(next, head);
// }


// const linkedListValues = (head, arr=[]) => {
//     if(head === null) return [];
//     arr = [head.val, ...linkedListValues(head.next)]
//     return arr;
// }


// let arr1 = linkedListValues(a);
// console.log(`Initial linked list: ${arr1}`);

// let newHead = reverseList(a);

// let arr2 = linkedListValues(newHead);
// console.log(`Linked list after reversal: ${arr2}`);


// Complexity:
// Iterative
// n = number of nodes
// Time: O(n)
// Space: O(1)

// Recursive
// n = number of nodes
// Time: O(n)
// Space: O(n)

File: 5. LinkedLists/5. reverse_list/reverse_list.py
class Node:
  def __init__(self, val):
    self.val = val
    self.next = None

a = Node("a")
b = Node("b")
c = Node("c")
d = Node("d")
e = Node("e")
f = Node("f")

a.next = b
b.next = c
c.next = d
d.next = e
e.next = f

# a -> b -> c -> d -> e -> f

def print_list(head):
  output = ""
  while(head.next is not None):
    output += str(head.val) + "->"
    head = head.next
  output += str(head.val)
  print(output)

# Iterative
def reverse_list(head):
  prev = None
  current = head
  while(current is not None):
    nextNode = current.next
    current.next = prev
    prev = current
    current = nextNode
  return prev

# Recursive
def reverse_list1(head, prev = None):
  if head is None:
    return prev
  nextNode = head.next
  head.next = prev
  return reverse_list1(nextNode, head)

print_list(a)
print_list(reverse_list(a)) # f -> e -> d -> c -> b -> a
# print_list(reverse_list1(a))